Add repeat function

osense opened the pull request:
This is a shortcut for a simple list comprehension, which can come in handy at times. For example, defining integer exponentiation with reduce and repeat:
```
> (defun exp (n k) (reduce #'*/2 (repeat k n)))
exp
> (exp 2 6)
64
```
The syntax is (repeat how-many what), which is like it is in Clojure. Personally, I'd prefer (repeat what how-many), however, since this is how Clojure has it, I guess it's better to keep it like this, not to confuse people?

I'm also planning on submitting pull requests for some other Clojure-inspired functions, like repeatedly and iterate. The idea here is that since someone spent a considerable amount of time designing Clojure, we can just steal the bits that we like :) And I think these are functions that many people might end up implementing themselves anyway.

oubiwann commented:
I'm a huge fan of Clojure, and if you're looked at some of the other bits in these includes, you've probably guessed that already ;-) (In fact, prior to LFE, Clojure was the language I had achieved the greatest amount of productivity, over every other language I've used.)

I'm definitely +1 in general for providing Clojure-like functions in LFE. We're getting more and more folks exploring LFE after Clojure, so it would be a nice welcoming :-) I've got some comments for you in this PR, but I'll put those in-line with the code.

Thanks for helping out!

oubiwann commented on include/core.lfe:
```diff
@@ -257,6 +257,12 @@
+;; Returns a list of n items, where each item is x.
+(defun repeat (n x)
+  (lc ((<- _ (seq n))) x))
+
```
First off, this is cool :-) I do love FP ...

but Erlang already has something like this. It would be more efficient to use it directly -- but we can still keep the ``repeat`` name:

```lisp
(defun repeat (n x)
  (lists:duplicate n x))
```

(See the docs [here](http://www.erlang.org/doc/man/lists.html#duplicate-2) for more details ...)

oubiwann replied:
You know, since it's so pretty, it might be worth putting a ``;;`` code comment before the function which preserves your implementation as education for the curious reader ...

osense replied:
You're right, I forgot about lists:duplicate/2. I added another commit that does what you suggested (but keeps the tests... if you want, I can remove those, seeing as it's practically just an alias now).

osense replied:
Interestingly enough, the lists:duplicate function uses a recursive call:
https://github.com/erlang/otp/blob/efbfe9602983ff451b864e557bdf3733222b78ba/lib/stdlib/src/lists.erl#L304 ,
which seems to take more time than a list comprehension to complete:
http://tryerl.seriyps.ru/#id=e4f2 .
Maybe we should submit a pull request to Erlang? :P

osense commented:
Alright, the definition is now replaced by a call to lists:duplicate, and I added some comments about why this is better than a list comprehension. Never mind what I said about LCs being faster before :)

osense commented:
I realized that we can have Clojure's repeat and repeatedly under one name, thanks to guards:
```
> (repeat 5 2)                    
(2 2 2 2 2)
> (repeat 2 #'random:uniform/0)   
(0.597447524783298 0.311326754804393)
```
so I added another commit that implements this.

One thing to note is that since I didn't add a special guard for the clause that triggers lists:duplicate, passing a negative integer and a function produces this error:
```
> (repeat -1 #'random:uniform/0)
exception error: function_clause
  in (: lists duplicate -1 #Fun<lfe_eval.23.2018457>)
```
i.e. since the first clause doesn't match, the arguments are passed straight to lists:duplicate, which has it's own guard clause to check for the positive n argument. Not sure if dodging this double check is worth the slightly-cryptic error.

oubiwann commented on include/core.lfe:
```diff
@@ -257,6 +257,31 @@
 
+;; Repeat
+;;
+;; Alias for lists:duplicate/2, but can also take a function as an argument.
```
Nice addition! (a function as an argument).

oubiwann commented:
Thanks, man -- I like it!
